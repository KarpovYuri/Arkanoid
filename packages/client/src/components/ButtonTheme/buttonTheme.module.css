
/* Dark mode btn */

.darkModeBtn {
	order: 9;
	position: relative;
	display: flex;
	justify-content: space-between;
	width: 51px;
	height: 26px;
	padding: 4px;
	border-radius: 50px;
	background-color: #272727;
}

.darkModeBtn::before {
	content: '';

	position: absolute;
	top: 0px;
	left: 0px;
	display: block;
	width: 24px;
	height: 24px;
	border-radius: 50%;
	background-color: #fff;
	transition: left 0.2s ease-in;
}

.darkModeBtnActive::before {
	left: 26px;
}

.darkModeBtnIcon {
	position: relative;
	z-index: 9;
}
