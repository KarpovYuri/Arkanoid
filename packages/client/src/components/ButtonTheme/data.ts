import sun from '../../assets/icons/sun.svg';
import moon from '../../assets/icons/moon.svg';
import rainbow from '../../assets/icons/rainbow.svg';
import { TThemes } from '../../typings';

export const THEMES: TThemes[] = [
  {
    theme_id: 1,
    theme_name: 'light',
    theme_icon: sun,
  },
  {
    theme_id: 2,
    theme_name: 'dark',
    theme_icon: moon,
  },
  {
    theme_id: 3,
    theme_name: 'rainbow',
    theme_icon: rainbow,
  },
];
