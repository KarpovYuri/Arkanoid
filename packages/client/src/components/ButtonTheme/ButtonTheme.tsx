import { useEffect, useState, useCallback } from 'react';
import classNames from 'classnames';
import styles from './ButtonTheme.module.css';
import { THEMES } from './data';
import { useAppSelector } from '../../store/hooks';
import { selectUserData } from '../../store/selectors';
import { UserData } from '../../store/typings';
import { themeAPI } from '../../api/ThemeAPI/ThemeAPI';
import { isSSR } from '../../utils/isSSR';

/**
 * README!!!
 * Для добавления новой цветовой темы необходимо:
 * 1. Создать новую тему в массиве THEMES в .\data.ts, подобрать иконку;
 * 2. Прописать цвета в client\src\styles\properties.css
 * 3. Тема в БД попадет автоматически
 */

export function ButtonTheme() {
  const userData: UserData | null = useAppSelector(selectUserData);
  const userId: number | undefined = userData?.id;

  // тема по умолчанию - первая, светлая
  const defaultTheme = THEMES[0];
  const defaultThemeId = defaultTheme.theme_id;

  const [theme, setTheme] = useState<number>(defaultThemeId);

  // ищем нужный атрибут по id и добавляем к документу (для применения стилей)
  const setThemeInDocument = useCallback((id: number) => {
    const currentTheme = THEMES.find((item) => item.theme_id === id);
    const currentThemeName: string = currentTheme
      ? currentTheme.theme_name : defaultTheme.theme_name;
    document.documentElement.setAttribute('data-theme', currentThemeName);
  }, [defaultTheme]);

  useEffect(() => {
    // говорят, в SSR local storage не алё
    const themeFromLS = localStorage.getItem('user-theme');

    /**
     * Пытаемся достать тему из LS, при этом не получив всяких undefined
     * Затем пробуем взять её из БД
     * Также нужна заглушка для установки дефолтной темы
     */
    if (themeFromLS && typeof JSON.parse(themeFromLS).theme_id !== 'undefined') {
      setTheme(JSON.parse(themeFromLS).theme_id);
      setThemeInDocument(JSON.parse(themeFromLS).theme_id);
    } else if (userId) {
      const themeFromDB = themeAPI.getUserTheme(userId);
      themeFromDB.then((res) => {
        if (res) {
          setTheme(res.theme_id);
          setThemeInDocument(res.theme_id);
        } else {
          setTheme(defaultThemeId);
          setThemeInDocument(defaultThemeId);
        }
      });
    } else {
      setTheme(defaultThemeId);
      setThemeInDocument(defaultThemeId);
    }
  }, [setTheme, defaultThemeId, userId, setThemeInDocument]);

  const handleThemeClick = (e: React.SyntheticEvent<EventTarget>) => {
    if (!(e.target instanceof HTMLButtonElement)) {
      return;
    }
    const selectedTheme = e.target.dataset.themeId;

    e.target.dataset.active = 'true';

    if (selectedTheme) {
      const selectedThemeNumeric = Number(selectedTheme);

      setThemeInDocument(selectedThemeNumeric);
      setTheme(selectedThemeNumeric);

      // говорят, в SSR local storage не алё
      if (!isSSR()) {
        localStorage.setItem('user-theme', JSON.stringify(selectedThemeNumeric));
      }

      // назначаем пользователю тему в БД
      if (userId) {
        themeAPI.setActiveUserTheme({ userId, themeId: selectedThemeNumeric });
      }
    }
  };

  const themeButtons = THEMES.map((item) => (
    <button
      type="button"
      aria-label="Set theme"
      onClick={(e) => handleThemeClick(e)}
      data-testid="toggleThemeBtn"
      data-theme-id={item.theme_id}
      key={item.theme_id + item.theme_name}
      data-theme-name={item.theme_name}
      className={classNames(styles[`btn-icon-${item.theme_name}`], styles.btnIcon)}
      data-active={theme === item.theme_id}
      style={{
        backgroundImage: `url("${item.theme_icon}")`,
      }}
    />
  ));

  return (
    <div className={styles.toggleThemeBtn} data-current={theme}>
      {themeButtons}
    </div>
  );
}
