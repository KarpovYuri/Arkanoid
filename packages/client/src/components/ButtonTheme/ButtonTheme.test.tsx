import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import '../../../mocks/matchMedia.mock'; // Must be imported before the tested file
import { ButtonTheme } from './ButtonTheme';

describe('ButtonTheme component tests', () => {
  test('Correct render', () => {
    render(<ButtonTheme />);
    const component = screen.getByTestId('toggleThemeBtn');

    expect(component).toBeInTheDocument();
    expect(component).toMatchSnapshot();
  });

  test('Change className onClick', () => {
    render(<ButtonTheme />);
    const component = screen.getByTestId('toggleThemeBtn');
    const classListBeforeClick = component.classList.length;

    fireEvent.click(component);

    const classListAfterClick = component.classList.length;

    expect(classListBeforeClick).not.toBe(classListAfterClick);
  });
});
