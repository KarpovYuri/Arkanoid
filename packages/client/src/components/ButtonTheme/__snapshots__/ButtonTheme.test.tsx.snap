// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonTheme component tests Correct render 1`] = `
<button
  class="darkModeBtn"
  data-testid="toggleThemeBtn"
  type="button"
>
  <img
    alt="Light mode"
    class="darkModeBtnIcon"
    src="default"
  />
  <img
    alt="Dark mode"
    class="darkModeBtnIcon"
    src="default"
  />
</button>
`;
